lib/wibox/widget/base.lua

```diff
@@ -321,6 +330,133 @@ function base.place_widget_at(widget, x, y, width, height)
     return base.place_widget_via_matrix(widget, matrix.create_translate(x, y), width, height)
 end
 
+-- Read the table, separate attributes from widgets
+local function parse_table(t, leave_empty)
+    local keys= {}
+    local max = 0
+    local attributes, widgets = {}, {}
+    for k,v in pairs(t) do
+        if type(k) == "number" then
+            -- As `ipairs` doesn't always work on sparse tables, update the
+            -- maximum
+            if k > max then
+                max = k
+            end
+
+            widgets[k] = v
+        else
+            attributes[k] = v
+        end
+    end
+
+    -- Pack the sparse table if the container doesn't support sparse tables
+    if not leave_empty then
+        widgets = util.table.from_sparse(widgets)
+        max = #widgets
+    end
+
+    return max, attributes, widgets
+end
+
+-- Recursively build a container from a declarative table
+local function drill(ids, content)
+    if not content then return end
+
+    -- Alias `widget` to `layout` as they are handled the same way
+    content.layout = content.layout or content.widget
+
+    -- Make sure  is not called on a function
+    local layout = type(content.layout) == "function" and  content.layout() or content.layout
+
+    -- Create layouts based on metatable __call
+    local l = layout.is_widget and layout or layout()
+
+    -- Get the number of children widgets (including nil widgets)
+    local max, attributes, widgets = parse_table(content, l.allow_empty_widget)
+
+    -- Get the optional identifier to create a virtual widget tree to place
+    -- in an "access table" to be able to retrieve the widget
+    local id = attributes.id
+
+    -- Clear the internal attributes
+    attributes.id, attributes.layout, attributes.widget = nil, nil, nil
+
+    for k = 1, max do
+        -- ipairs cannot be used on sparse tables
+        local v, id2, e = widgets[k], id, nil
+        if v then
+            -- It is another declarative container, parse it
+            if not v.is_widget then
+                e, id2 = drill(ids, v)
+                widgets[k] = e
+            end
+
+            -- Place the widget in the access table
+            if id2 then
+                l  [id2] = e
+                ids[id2] = e
+                id2      = nil
```

Comment:
Why unset `id2`? That doesn't really matter.

Also, you are adding random keys to random widgets (the `l[id2] = e`-part). If I give a widget the id `add`, `draw` or something like that, stuff will break.
What's the point of this `id` stuff (I don't see it documented, I think). If users should use that, wouldn't it be enough to just return the `ids` table (that's already done now) and people have to save that somewhere themselves?

Oh and all of this might break if a widget is shown in more than one place, but I guess I can live with that (it needs something made on purpose to break it)....